File: android/app/src/main/java/com/openavc/panel/kiosk/KioskPreferences.kt
package com.openavc.panel.kiosk

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * Persists kiosk configuration: whether the user has asked the app to lock
 * itself into kiosk mode, and a salted SHA-256 hash of the admin PIN.
 *
 * The PIN is never stored in plaintext. It only needs to resist shoulder
 * surfing and casual tampering by end users at the panel — not a cryptographic
 * attacker with filesystem access — but using a salted hash is still cheap and
 * avoids leaking the PIN to anyone who dumps SharedPreferences.
 */
class KioskPreferences(context: Context) {

    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences(FILE_NAME, Context.MODE_PRIVATE)

    var kioskEnabled: Boolean
        get() = prefs.getBoolean(KEY_ENABLED, false)
        set(value) = prefs.edit { putBoolean(KEY_ENABLED, value) }

    fun hasPin(): Boolean = prefs.contains(KEY_PIN_HASH)

    fun setPin(pin: String) {
        val salt = ByteArray(16).also { SecureRandom().nextBytes(it) }
        val hash = hash(pin, salt)
        prefs.edit {
            putString(KEY_PIN_SALT, salt.toHex())
            putString(KEY_PIN_HASH, hash.toHex())
        }
    }

    fun clearPin() {
        prefs.edit {
            remove(KEY_PIN_SALT)
            remove(KEY_PIN_HASH)
        }
    }

    fun checkPin(pin: String): Boolean {
        val saltHex = prefs.getString(KEY_PIN_SALT, null) ?: return false
        val expected = prefs.getString(KEY_PIN_HASH, null) ?: return false
        val salt = saltHex.hexToBytes() ?: return false
        val actual = hash(pin, salt).toHex()
        return constantTimeEquals(expected, actual)
    }

    private fun hash(pin: String, salt: ByteArray): ByteArray {
        val md = MessageDigest.getInstance("SHA-256")
        md.update(salt)
        md.update(pin.toByteArray(Charsets.UTF_8))
        return md.digest()
    }

    private fun ByteArray.toHex(): String =
        joinToString(separator = "") { "%02x".format(it.toInt() and 0xff) }

    private fun String.hexToBytes(): ByteArray? {
        if (length % 2 != 0) return null
        return ByteArray(length / 2) { i ->
            val high = Character.digit(this[i * 2], 16)
            val low = Character.digit(this[i * 2 + 1], 16)
            if (high < 0 || low < 0) return null
            ((high shl 4) or low).toByte()
        }
    }

    private fun constantTimeEquals(a: String, b: String): Boolean {
        if (a.length != b.length) return false
        var result = 0
        for (i in a.indices) result = result or (a[i].code xor b[i].code)
        return result == 0
    }

    companion object {
        private const val FILE_NAME = "openavc_kiosk_prefs"
        private const val KEY_ENABLED = "kiosk_enabled"
        private const val KEY_PIN_HASH = "pin_hash"
        private const val KEY_PIN_SALT = "pin_salt"
    }
}

File: ios/OpenAVCPanel/Kiosk/KioskPreferences.swift
import CryptoKit
import Foundation
import Security

/// Persists the dedicated-panel settings: whether the panel should ask iOS to
/// lock it every time it opens, and a salted SHA-256 hash of the admin PIN.
///
/// Mirrors `KioskPreferences.kt` on Android, keys included, so the two are
/// legible side by side. `kioskEnabled` keeps the Android name; on iOS it means
/// "request a Guided Access session on launch", which only takes on an iPad an
/// MDM has allowed to lock itself (Autonomous Single App Mode).
///
/// The PIN is never stored in plaintext. It has to resist shoulder surfing and
/// casual tampering by people at the panel, not an attacker with the file
/// system, and a salted hash is cheap.
struct KioskPreferences {
    private let defaults: UserDefaults

    private static let suiteName = "openavc_kiosk_prefs"

    private enum Key {
        static let enabled = "kiosk_enabled"
        static let pinHash = "pin_hash"
        static let pinSalt = "pin_salt"
    }

    init(defaults: UserDefaults? = nil) {
        self.defaults = defaults ?? UserDefaults(suiteName: Self.suiteName) ?? .standard
    }

    var kioskEnabled: Bool {
        get { defaults.bool(forKey: Key.enabled) }
        nonmutating set { defaults.set(newValue, forKey: Key.enabled) }
    }

    func hasPin() -> Bool {
        defaults.string(forKey: Key.pinHash) != nil
    }

    func setPin(_ pin: String) {
        var salt = [UInt8](repeating: 0, count: 16)
        _ = SecRandomCopyBytes(kSecRandomDefault, salt.count, &salt)
        let saltData = Data(salt)
        defaults.set(Self.hex(saltData), forKey: Key.pinSalt)
        defaults.set(Self.hex(Self.hash(pin, salt: saltData)), forKey: Key.pinHash)
    }

    func clearPin() {
        defaults.removeObject(forKey: Key.pinSalt)
        defaults.removeObject(forKey: Key.pinHash)
    }

    func checkPin(_ pin: String) -> Bool {
        guard let saltHex = defaults.string(forKey: Key.pinSalt),
              let expected = defaults.string(forKey: Key.pinHash),
              let salt = Self.bytes(fromHex: saltHex)
        else { return false }
        let actual = Self.hex(Self.hash(pin, salt: salt))
        return Self.constantTimeEquals(expected, actual)
    }

    private static func hash(_ pin: String, salt: Data) -> Data {
        var hasher = SHA256()
        hasher.update(data: salt)
        hasher.update(data: Data(pin.utf8))
        return Data(hasher.finalize())
    }

    static func hex(_ data: Data) -> String {
        data.map { String(format: "%02x", $0) }.joined()
    }

    static func bytes(fromHex hex: String) -> Data? {
        guard hex.count % 2 == 0 else { return nil }
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let next = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<next], radix: 16) else { return nil }
            data.append(byte)
            index = next
        }
        return data
    }

    private static func constantTimeEquals(_ a: String, _ b: String) -> Bool {
        guard a.utf8.count == b.utf8.count else { return false }
        var result: UInt8 = 0
        for (x, y) in zip(a.utf8, b.utf8) { result |= x ^ y }
        return result == 0
    }
}
